client/components/site-verticals-suggestion-search/index.jsx

```diff
@@ -26,55 +26,126 @@ import './style.scss';
+			this.setSearchResults( this.props.verticals );
+		}
 	}
 
-	// When a user is keying through the results,
-	// only update the vertical when they select a result.
+	/**
+	 * Show the popular topics component when the search field is empty and
+	 * `props.showPopular` tell us to.
+	 *
+	 * @returns {Bool} Whether we should display the popular topics component.
+	 */
+	shouldShowPopularTopics = () => ! this.state.inputValue && this.props.showPopular;
+
+	/**
+	 * Checks for the existence of vertical results. If there are none we assume
+	 *
+	 * @returns {Bool} Whether we should display the popular topics component.
+	 */
+	isVerticalSearchPending = () => this.state.inputValue && 0 === this.props.verticals.length;
+
+	/**
+	 * Sets `state.results` with incoming vertical results, retaining previous non-user vertical search results
+	 * if the incoming vertical results contain only user-defined results.
+	 *
+	 * This function could better be performed in the backend eventually.
+	 *
+	 * @param {Array} results Incoming vertical results
+	 */
+	setSearchResults = results => {
+		if ( results && results.length ) {
+			const { candidateVerticals, inputValue, isSuggestionSelected } = this.state;
+			// If the user is typing
+			// and the only result is a user input (non-vertical),
+			// and we have some previous results
+			// then concat that with the previous results and remove the last user input
+			if (
+				! isSuggestionSelected &&
+				! find( results, item => ! item.isUserInputVertical ) &&
+				1 < candidateVerticals.length
+			) {
+				results = candidateVerticals.filter( item => ! item.isUserInputVertical ).concat( results );
+			}
+
+			this.setState( { candidateVerticals: results }, () =>
+				this.updateVerticalData( this.searchForVerticalMatches( inputValue ), inputValue )
+			);
+		}
+	};
+
+	getNewRailcar = () => ( {
+		id: `${ uuid().replace( /-/g, '' ) }-site-vertical-suggestion`,
+		fetch_algo: '/verticals',
+		action: 'site_vertical_selected',
+	} );
+
+	/**
+	 * Searches the API results for a direct match on the user search query.
+	 *
+	 * @param {String} value Search query array
+	 * @returns {Object?} An object from the vertical results array
+	 */
 	searchForVerticalMatches = ( value = '' ) =>
 		find(
 			this.props.verticals,
-			item => item.verticalName.toLowerCase() === value.toLowerCase() && !! item.preview
+			item => item.verticalName.toLowerCase() === value.toLowerCase().trim() && !! item.preview
 		);
 
-	// TODO: once the siteVertical state got simplified, this can be removed.
-	updateVerticalData = ( result, value ) =>
+	/**
+	 * Callback to be passed to consuming component when the search value is updated.
+	 * TODO: once the siteVertical state got simplified, this can be removed.
+	 *
+	 * @param {Object} verticalData An object from the vertical results array
+	 * @param {String} value Search query array
+	 */
+	updateVerticalData = ( verticalData, value = '' ) => {
+		value = value.trim();
```

Comment:
Oops. Let's avoid mutating the argument 😉 Creating a copy of it in another variable is fine.